Try: Restore block mover for floats.

jasmussen opened the pull request:
This PR hopes to fix #12736, #11424, #10300.

Currently we hide the block mover for any floated block. We do this because if you have a left floated image followed by a paragraph, the block mover for the image would overlap exactly with the block mover for the subsequent paragraph block. This would cause moving a float to be fiddly and virtually impossible as the hover states for the two conflicted.

This PR is an experiment to mitigate, or possibly fix that. It's a try branch because it is not necessarily a solid fix. What this PR does:

- It makes it so the block mover cannot be invoked when only hovering a float, but when the float is selected they are permanently visible.
- It changes some z-indexes so floats always have a higher z-index than not a float.
- It changes the footprint of the side-UI container to "cover up" the underlying paragraph block, so it won't invoke unlesss you move the mouse below the footprint.

This is best explained in some GIFs. Here's one where the block mover footprint is coral colored to help explain what's going on:

![float movers](https://user-images.githubusercontent.com/1204802/49729541-1563c280-fc76-11e8-9c31-db436bcb4514.gif)

Here's without that coral debug color:

![without debug](https://user-images.githubusercontent.com/1204802/49729558-1ac10d00-fc76-11e8-8a3f-c52ed76b4594.gif)

The reason this is a "Try" branch is that when you move the mouse over the adjacent paragraph, the block mover is show as overlapping the selected floats block mover. But due to the rearranging of z-indexes, at least this is only a visual issue. A click on the floats block mover will still work as intended.

Why can't the float have on-hover block movers like every other block, you ask? Picture again the test-case of a left floated image followed by a paragraph of text. In this case, hovering over the image would show a block mover that would visually appear to be that of the paragraph block. By showing it when the float is selected, the context is clear.

We could experiment with not showing the hover block mover for float-adjacent blocks when hovered, but this isn't feasible because you might have a float, and then a block that clears this float, in which case the rule would break down.

Lay your thoughts on me.

Commit:
Try: Restore block mover for floats.

This PR hopes to fix #12736, #11424, #10300.

Currently we hide the block mover for any floated block. We do this because if you have a left floated image followed by a paragraph, the block mover for the image would overlap exactly with the block mover for the subsequent paragraph block. This would cause moving a float to be fiddly and virtually impossible as the hover states for the two conflicted.

This PR is an experiment to mitigate, or possibly fix that. It's a try branch because it is not necessarily a solid fix. What this PR does:

- It makes it so the block mover cannot be invoked when only hovering a float, but when the float is selected they are permanently visible.
- It changes some z-indexes so floats always have a higher z-index than not a float.
- It changes the footprint of the side-UI container to "cover up" the underlying paragraph block, so it won't invoke unlesss you move the mouse below the footprint.

This is best explained in some GIFs.

The reason this is a "Try" branch is that when you move the mouse over the adjacent paragraph, the block mover is show as overlapping the selected floats block mover. But due to the rearranging of z-indexes, at least this is only a visual issue. A click on the floats block mover will still work as intended.

Why can't the float have on-hover block movers like every other block, you ask? Picture again the test-case of a left floated image followed by a paragraph of text. In this case, hovering over the image would show a block mover that would visually appear to be that of the paragraph block. By showing it when the float is selected, the context is clear.

We could experiment with not showing the hover block mover for float-adjacent blocks when hovered, but this isn't feasible because you might have a float, and then a block that clears this float, in which case the rule would break down.

Lay your thoughts on me.

jasmussen commented on packages/editor/src/components/block-list/block.js:
```diff
@@ -553,25 +553,25 @@ export class BlockListBlock extends Component {
 								<BlockMultiControls rootClientId={ rootClientId } />
 							) }
 							<div className="editor-block-list__block-edit">
+								{ shouldRenderMovers && (
```
For the block mover to be able to work for a right-floated block, it has to be inside `.editor-block-list__block-edit`, or the control will show up way on the left of the main column. Same case as when we refactored the floats, essentially.

jasmussen commented on packages/editor/src/components/block-list/style.scss:
```diff
@@ -1,8 +1,8 @@
 .editor-block-list__layout .components-draggable__clone {
-	// Hide the Block UI when dragging the block
-	// This ensures the page scroll properly (no sticky elements)
+	// Hide the Block UI when dragging the block.
```
Just a few comment cleanups while I'm in here.

chrisvanpatten commented:
I think the iteration could happen in a separate PR but I think this is another case where some version of #7211 would help with the mover overlap issues.

jasmussen commented:
> I think the iteration could happen in a separate PR but I think this is another case where some version of #7211 would help with the mover overlap issues.

I know I've been skeptical in the past, but I'm starting to agree especially for nested blocks and now floats. I still don't really see the value it brings to top level blocks, except perhaps the consistency it would bring, but for those blocks which for now are still the most common, the extra chrome behind the controls just seems heavier.

chrisvanpatten commented:
@jasmussen Yeah it's primarily a consistency thing, although I do think it will be more practically valuable as the editor starts to reflect the front-end more and more and the "base level" editor canvas isn't so bare. But that could be a future iteration; it could definitely focus just on InnerBlocks for the moment.

jasmussen commented:
@youknowriad For now I've set this in "future/phase 2". But I would appreciate a pair of eyes if you have time — could we get this into a plugin release like 4.9? I'd love for it to land early between major releases.

youknowriad commented:
I've put this PR in 4.9 to get some review 👀 It doesn't mean it's going to ship there though. Depends on the reviews.

jasmussen commented:
I'm not necessarily super antsy that it land in 4.9 or even 5.0, but I appreciate it being on the radar! Thanks Riad! 💕

ZebulanStanphill commented on packages/editor/src/components/block-list/style.scss:
```diff
@@ -1,8 +1,8 @@
+	// This ensures the page scroll properly (no sticky elements).
 	.editor-block-contextual-toolbar {
-		// I think important is fine here to avoid over complexing the selector
+		// I think important is fine here to avoid over complexing the selector.
```
```suggestion
		// I think important is fine here to avoid over-complicating the selector.
```

ZebulanStanphill commented on packages/editor/src/components/block-list/style.scss:
```diff
@@ -168,7 +168,7 @@
 
-	// selection style for multiple blocks
+	// Selection style for multiple blocks.
 	&.is-multi-selected *::selection {
```
```suggestion
	&.is-multi-selected ::selection {
```
I think that asterisk is redundant.

jasmussen replied:
Good thought, but I'd prefer not actually making this change in this PR. It is already sort of big surgery and I'd like to minimize the changes as much as possible. Given that code did not change in this PR, only the comment, I'd prefer to look at this separately.

Commit:
Address feedback, and improve dragging edgecase.

This addresses comment feedback by @ZebulanStanphill, thanks, and it also hides the block mover from the ghost when you are dragging.

jasmussen commented:
Rebased, and addressed feedback. Tested again. I think this is good to go, it seems to work well for me. It would be good to get in _after 5.1_, but before 5.2, so I think now might be a good time to merge this.

One thing I fixed was hiding the block mover from the gray ghost when dragging:

![dragging](https://user-images.githubusercontent.com/1204802/51970393-14316380-2477-11e9-89d4-dcb1c211c524.gif)

gziolo commented:
I only noticed one inconsistency with other blocks. Movers show up even when I move mouse close to the right edge of the block. It behaves differently with blocks that span the width of the editor:

![screen shot 2019-02-04 at 16 10 17](https://user-images.githubusercontent.com/699132/52216898-96ff5780-2897-11e9-82fe-ef3cf3d14036.png)

gziolo reviewed (approved):
Nice improvement 👍

jasmussen commented:
> I only noticed one inconsistency with other blocks. Movers show up even when I move mouse close to the right edge of the block. It behaves differently with blocks that span the width of the editor:

Yes, this is intentional. The hover effect is gone from floats, and the movers are now permanently visible when the block is selected.

When I worked on this, I recall it feeling best this way, notably in situations where you have a right floated image, then a text block that wraps around that, and then have the text block seleted. I can take another glance and see if it's possible to keep the old behavior and have a good experience, but I recall this change being very intentional and part of enabling it for floats in the first place (even if the exact reason now escapes me).

ZebulanStanphill commented:
@kjellr In that second GIF, it seems like those movement control buttons should always have a solid background.

jasmussen commented:
Rebased, to keep this fresh. 

It seems like this is mostly ready to go, the remaining sticking point being what Kjell and ZebulanStanphill brought up, about overlap.

Is the following the solution?

![bgs](https://user-images.githubusercontent.com/1204802/52260145-7978cf00-2925-11e9-8c59-1d9ffa695608.gif)

If yes, let me know and I can push that as a fix and then merge.

kjellr commented:
> Is the following the solution?

Sure, that works for me. We can push that in and keep iterating. 👍

ZebulanStanphill commented:
@jasmussen That change looks good to me.

mapk commented:
The improvements are great, but the border on each of the positioning elements does leave a desire for iteration. 😉 

I tried to see if we can just add the bordered state to the arrows only (not sure if it helps).

![screen shot 2019-02-05 at 5 00 39 pm](https://user-images.githubusercontent.com/617986/52314597-43107380-2968-11e9-8003-7e307d80c146.png)

Let's not let this hold up the PR though. Thanks for the work, Joen!

Commit:
Add borders around floats.

jasmussen commented:
Added the borders around the movers for floats, as proposed. 

> I tried to see if we can just add the bordered state to the arrows only (not sure if it helps).

I do agree that the movers could be visually enhanced a fair bit — perhaps just making it feel like a "pill button with separators" instead of 3 stacked rounded buttons might help? The background is important even for the drag handle when it overlaps colored backgrounds or images. 

But right now this uses the same CSS as any nested blocks, which also have the bordered movers. Let's explore improving those styles separately. I know @chrisvanpatten has ideas around that also.

jasmussen commented:
@gziolo Unless you object, I think we should land this once the checks pass!

gziolo commented:
Merged 🎉

jasmussen commented:
`itshappening.rmvb`!

senadir commented:
Hi, this fix seems to cause a new bug #15923, #15930
setting the image block z-index to 81 overrides the block toolbar 
this pull request should fix it #15966 while keeping everything right
